pkg/builder/endpoints.go

```diff
@@ -0,0 +1,262 @@
+package builder
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/common/hash"
+)
+
+const (
+	IPAddressType       v1alpha1.AddressType = "IPAddress"
+	HostnameAddressType v1alpha1.AddressType = "Hostname"
+	// HostnameRegex checks for at least two groups of allowed for URLs characters separated by "."
+	HostnameRegex        = "^(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)\\.(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)$"
+	WildcardGeo   string = "*"
+
+	DefaultTTL      = 60
+	DefaultCnameTTL = 300
+	IDLength        = 6
+)
+
+// Target wraps a kubernetes ingress traffic resource e.g.Gateway, Ingress, Route etc.. but can wrap any resources
+// that has the desired geo and weight labels being applied, and can provide the required target addresses data.
+// This should be implemented as required for each type of ingress resource i.e. Gateway
+type Target interface {
+	metav1.Object
+	GetAddresses() []v1alpha1.TargetAddress
+}
+
+// EndpointsBuilder builds an endpoints array.
+type EndpointsBuilder struct {
+	// target kubernetes resource that may have geo/weight labels applied and provides target addresses.
+	target Target
+
+	// hostname to be used for creation of endpoints. There could be multiple hostname values for a
+	// single target. This builder delegates burden of determining valid hostnames and managing
+	// an array of endpoints for each of hostname values to the consumer of this API
+	hostname string
+
+	// loadBalancing specification (Optional),
+	// If set the builder will create a loadbalanced set of endpoints for the target resource.
+	// If unset, the builder will create a simple set of endpoints for the target resource.
+	loadBalancing *v1alpha1.LoadBalancingSpec
+
+	// id is a way to distinguish endpoints created for the same target
+	// with the same hostname for a different cluster (Optional)
+	id string
+}
+
+// NewEndpointsBuilder returns a new endpoints builder
+func NewEndpointsBuilder(target Target, hostname string) *EndpointsBuilder {
+	return &EndpointsBuilder{
+		target:   target,
+		hostname: hostname,
+	}
+}
+
+// WithLoadBalancing provides builder with needed parameters to generate load-balancing set of endpoints.
+// If not used the builder will provide a simple set of endpoints
+func (builder *EndpointsBuilder) WithLoadBalancing(id string, loadbalancing *v1alpha1.LoadBalancingSpec) *EndpointsBuilder {
+	builder.id = id
+	builder.loadBalancing = loadbalancing
+	return builder
+}
```

Comment:
So with the changes that have been made to [simplify the DNSPolicy API](https://github.com/Kuadrant/kuadrant-operator/pull/858), and the fact that we **don't** intend to define the LoadBalancingSpec here and reference it in the DNSPolicy, the inputs to this builder feel like they could also just be made to fit how we intend it to be used. 

We have no real need for a target resource now since we don't need to inspect labels or anything from a k8s resource to determine weights and geos, so the `Target` interface doesn't seem like it's that useful now.
Since this API is de-coupled from the policy LoadBalancingSpec, we can just make the input for the loadbalanced endpoint generation the things required by `getLoadBalancedEndpoints`. 

Something along these lines might be better now, wdyt?:
```.go
type AddressType string

type TargetAddress struct {
	Type  AddressType
	Value string
}

type LoadBalancing struct {
	id         string
	weight     int
	geo        string
	defaultGeo bool
}

// EndpointsBuilder builds an endpoints array.
type EndpointsBuilder struct {
	hostname      string
	loadBalancing *LoadBalancing
	addresses     []TargetAddress
}

// NewEndpointsBuilder returns a new endpoints builder
func NewEndpointsBuilder(hostname string) *EndpointsBuilder {
	return &EndpointsBuilder{
		hostname: hostname,
	}
}

func (builder *EndpointsBuilder) WithLoadBalancing(lb *LoadBalancing) *EndpointsBuilder {
	builder.loadBalancing = lb
	return builder
}

func (builder *EndpointsBuilder) WithLoadBalancingFor(id string, geo string, weight int, defaultGeo bool) *EndpointsBuilder {
	return builder.WithLoadBalancing(&LoadBalancing{
		id:         id,
		weight:     weight,
		geo:        geo,
		defaultGeo: defaultGeo,
	})
}

func (builder *EndpointsBuilder) WithAddresses(addresses []TargetAddress) *EndpointsBuilder {
	builder.addresses = addresses
	return builder
}
```

Reply:
This is also kind of similar to builder for dnspolicy https://github.com/Kuadrant/kuadrant-operator/blob/main/api/v1alpha1/dnspolicy_types.go#L277-L283

Reply:
I'd still want to stick with only the following methods on the builder: 
- `NewEndpointsBuilder()` 
- `WithLoadBalancing()`
- `Build` 
To highlight that there is only one option while generating endpoints: use lb or not. In other words I'd like to not have `WithAddresses` as they are mandatory for any endpoint. 
But you are correct - there is a little use for the `LoadbalancingSpec` struct. I'd keep it anyway as it makes validation more convenient, and will borrow your `WithLoadBalancingFor`. The later should make it easier for people to use this API. 

I'll still keep the `Target` interface as we use the name and namespace to generate unique shortcode for load-balanced endpoint. 
And I really like the ide to be more consistent with patterns we have in the kuadrant-operator

Reply:
Well, you have it currently as target and hostname are mandatory and required in the constructor, you could do the same with addresses if you really wanted for consistency. The only "mandatory" things are hostname and addresses. loadbalancing is optional but requires if set, id, name and namesapce all of which woudl make more sense in the loadbalancing struct IMO.